src/devices/PiJuice/PiJuice.cs

```diff
@@ -0,0 +1,240 @@
+        public void SetIdEepromAddress(IdEepromAddress epromAddress)
+        {
+            WriteCommandVerify(PiJuiceCommand.IdEepromAddress, new byte[] { (byte)epromAddress });
+        }
+
+        /// <summary>
+        /// Write a PiJuice command
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="data">The data to write</param>
+        internal void WriteCommand(PiJuiceCommand command, byte[] data)
+        {
+            byte tries = 0;
+            IOException innerEx = null;
+            // When writing/reading to the I2C port, PiJuice doesn't respond on time in some cases
+            // So we wait a little bit before retrying
+            // In most cases, the I2C read/write can go thru without waiting
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    // create request
+                    var buffer = new byte[data.Length + 2];
+                    Array.Copy(data, 0, buffer, 1, data.Length);
+                    buffer[0] = (byte)command;
+                    buffer[buffer.Length - 1] = GetCheckSum(data, all: true);
+                    _i2cDevice.Write(buffer);
+                    return;
+                }
+                catch (IOException ex)
+                {
+                    // Give it another try
+                    innerEx = ex;
+                    tries++;
+                    DelayHelper.DelayMilliseconds(ReadRetryDelay, false);
+                }
+            }
+
+            throw new IOException($"{nameof(WriteCommand)}: Failed to write command {command}", innerEx);
+        }
+
+        /// <summary>
+        /// Write a PiJuice command and verify
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="data">The data to write</param>
+        /// <param name="delay">The delay before reading the data</param>
+        internal void WriteCommandVerify(PiJuiceCommand command, byte[] data, int delay = 0)
+        {
+            WriteCommand(command, data);
+
+            if (delay > 0)
+            {
+                DelayHelper.DelayMilliseconds(delay, false);
+            }
+
+            var response = ReadCommand(command, (byte)data.Length);
+
+            if (!data.SequenceEqual(response))
+            {
+                throw new IOException($"{nameof(WriteCommandVerify)}: Failed to write command {command}");
+            }
+        }
+
+        /// <summary>
+        /// Read data from PiJuice
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="length">The length of the data to be read</param>
+        /// <returns>Returns an array of the bytes read</returns>
+        internal byte[] ReadCommand(PiJuiceCommand command, byte length)
+        {
+            byte[] outArray = new byte[length + 1];
+            byte tries = 0;
+            IOException innerEx = null;
```

Comment:
why don't you just define it where it's raised?

Reply:
It is inside a try catch which is itself inside a while loop for a specific number of retries. Also this code is based on the GrovePi.

I could change the catch to check if the attempts has exceeded the retries and change the while loop to a for loop?

Reply:
Yes, I would raise the exception in the catch part if the tries are higher than the maximum. But that's ok as well like this. It was really to avoid allocating the IOException for nothing